File: cpp/src/ir_builder.cpp
/*******************************************************************************
 *  Function Definition for translation functions.                             *
 ******************************************************************************/

#include "c_ast.hpp"
#include "koopa_ast.hpp"
#include <stdexcept>

std::unique_ptr<koopa_ast::Program>
translate_comp_unit_c_ast(const c_ast::CompUnitAST &);
std::unique_ptr<koopa_ast::Function>
translate_func_def_c_ast(const c_ast::FuncDefAST &);
std::unique_ptr<koopa_ast::Type>
translate_func_type_c_ast(const c_ast::FuncTypeAST &);
std::unique_ptr<koopa_ast::BasicBlock>
translate_block_c_ast(const c_ast::BlockAST &block, std::string = "");
koopa_ast::Value *translate_stmt_c_ast(const c_ast::StmtAST &,
                                       koopa_ast::BasicBlock &);
koopa_ast::Value *translate_exp_c_ast(const c_ast::ExpAST &,
                                      koopa_ast::BasicBlock &);
koopa_ast::Value *translate_primary_exp_c_ast(const c_ast::PrimaryAST &,
                                              koopa_ast::BasicBlock &);
koopa_ast::Value *translate_unary_exp_c_ast(const c_ast::UnaryExpAST &,
                                            koopa_ast::BasicBlock &);
koopa_ast::Integer *translate_number_c_ast(const c_ast::NumberAST &,
                                           koopa_ast::BasicBlock &);

/*******************************************************************************
 *  Implementation Details for going from each C AST nodes to Koopa Node.      *
 *******************************************************************************/

/**
 * Going from NumberAST to koopa Integer
 */
// std::unique_ptr<koopa_ast::Integer>
koopa_ast::Integer *translate_number_c_ast(const c_ast::NumberAST &number,
                                           koopa_ast::BasicBlock &block) {
  auto *ret = block.Make<koopa_ast::Integer>(false, number.int_val);
  return ret;
}

koopa_ast::Value *translate_primary_exp_c_ast(const c_ast::PrimaryAST &primary,
                                              koopa_ast::BasicBlock &block) {
  if (auto *p = dynamic_cast<const c_ast::PrimaryASTExp *>(&primary)) {
    auto *exp = dynamic_cast<const c_ast::ExpAST *>(p->exp.get());
    if (!exp)
      throw std::runtime_error("ir_builder error: PrimaryASTExp expects "
                               "ExpAST at param `exp`");

    return translate_exp_c_ast(*exp, block);
  } else if (auto *p_num =
                 dynamic_cast<const c_ast::PrimaryASTNumber *>(&primary)) {
    auto *num = dynamic_cast<const c_ast::NumberAST *>(p_num->number.get());
    if (!num)
      throw std::runtime_error(
          "ir_builder error: PrimaryASTNumber expects NumberAST at param "
          "`number`");
    return block.Make<koopa_ast::Integer>(false, num->int_val);
  } else {
    throw std::runtime_error(
        "ir_builder error: PrimaryAST must have one of the following "
        "implementation: {PrimaryASTExp, PrimaryASTNumber} ");
  }
}

koopa_ast::Value *translate_unary_exp_c_ast(const c_ast::UnaryExpAST &unary,
                                            koopa_ast::BasicBlock &block) {

  if (auto *p = dynamic_cast<const c_ast::UnaryExpASTPrimary *>(&unary)) {

    auto *prim = dynamic_cast<const c_ast::PrimaryAST *>(p->primary_exp.get());
    if (!prim)
      throw std::runtime_error("ir_builder error: UnaryExpASTPrimary expects "
                               "PrimaryAST at param `primary_exp`");
    return translate_primary_exp_c_ast(*prim, block);
  } else if (auto *op =
                 dynamic_cast<const c_ast::UnaryExpASTOpUnary *>(&unary)) {

    auto *u_exp = dynamic_cast<const c_ast::UnaryExpAST *>(op->unary_exp.get());

    if (!u_exp)
      throw std::runtime_error("ir_builder error: UnaryExpASTOpUnary expects "
                               "UnaryExpAST at param `unary_exp`");

    auto *u_exp_ast = translate_unary_exp_c_ast(*u_exp, block);

    // Generate the instruction based on the operation type
    switch (op->unary_op) {
    case c_ast::UnaryOp::PLUS: {
      return u_exp_ast;
      break;
    }
    case c_ast::UnaryOp::MINUS: {
      auto *zero = block.Make<koopa_ast::Integer>(false, 0);
      auto *inst = block.Make<koopa_ast::Binary>(true, koopa_ast::BinaryOp::Sub,
                                                 zero, u_exp_ast);
      return inst;
      break;
    }
    case c_ast::UnaryOp::BANG: {
      auto *zero = block.Make<koopa_ast::Integer>(false, 0);
      auto *inst = block.Make<koopa_ast::Binary>(true, koopa_ast::BinaryOp::Eq,
                                                 u_exp_ast, zero);
      return inst;
      break;
    }
    case c_ast::UnaryOp::TILDE: {
      auto *all_one = block.Make<koopa_ast::Integer>(false, -1);
      auto *inst = block.Make<koopa_ast::Binary>(true, koopa_ast::BinaryOp::Xor,
                                                 u_exp_ast, all_one);
      return inst;
      break;
    }
    }
  } else {
    throw std::runtime_error(
        "ir_builder error: UnaryExpAST must have one of the following "
        "implementation: {UnaryExpASTPrimary, UnaryExpASTOpUnary} ");
  }
}

koopa_ast::Value *translate_exp_c_ast(const c_ast::ExpAST &exp,
                                      koopa_ast::BasicBlock &block) {
  auto *unary_exp =
      dynamic_cast<const c_ast::UnaryExpAST *>(exp.unary_exp.get());
  if (!unary_exp)
    throw std::runtime_error(
        "ir_builder error: ExpAST expects UnaryExpAST at param `unary_exp`");
  return translate_unary_exp_c_ast(*unary_exp, block);
}

/**
 * Going from StmtAST to koopa stmt
 *
 */
koopa_ast::Value *translate_stmt_c_ast(const c_ast::StmtAST &stmt,
                                       koopa_ast::BasicBlock &block) {
  // Cast and check that it has the correct type
  auto *exp = dynamic_cast<const c_ast::ExpAST *>(stmt.exp.get());
  if (!exp)
    throw std::runtime_error(
        "ir_builder error: StmtAST expects ExpAST at param `exp`");

  // Translate the expression inside the return
  auto *val = translate_exp_c_ast(*exp, block);

  return block.Make<koopa_ast::Return>(true, val);
}

/**
 * Converting FuncType C AST to just Type koopa IR reps.
 *
 * HACK: Currently since we don't have type for the c_ast,
 * this function will always return i32.
 */
std::unique_ptr<koopa_ast::Type>
translate_func_type_c_ast(const c_ast::FuncTypeAST &func_type) {
  // HACK: Currently it is just hard matching as it has to be int
  return std::make_unique<koopa_ast::Type>(koopa_ast::Type::I32());
}

/**
 * Converting a Block C AST to koopa Block for now.
 *
 * NOTE: The meaning of a block in C and a block in koopa is different.
 */
std::unique_ptr<koopa_ast::BasicBlock>
translate_block_c_ast(const c_ast::BlockAST &block, std::string name) {
  auto ret = std::make_unique<koopa_ast::BasicBlock>(name);

  auto *stmt = dynamic_cast<const c_ast::StmtAST *>(block.stmt.get());
  if (!stmt)
    throw std::runtime_error(
        "ir_builder error: BlockAST expects to have StmtAST at param `stmt`");

  translate_stmt_c_ast(*stmt, *ret.get());

  return ret;
}

/**
 * Converting a FuncDefAST to a koopa function as the following:
 *
 *
 * ```
 *
 *  c_ast::FuncDefAST:name (string)
 *    -> koopa_ast::Function:name (string)
 *
 *  c_ast::FuncDefAST:type (string)
 *    -> koopa_ast::Function:type (koopa_ast::Type)
 *
 *  c_ast::FuncDefAST:block (c_ast::BlockAST)
 *    -> koopa_ast::Function:basicblocks (vector<koopa_ast::BasicBlock>)
 *
 * ```
 *
 *
 * NOTE: I think this does satisfy the definition of a koopa block as it does
 * only have one point of entry and only leave or return. However, this may not
 * be the most optimised way to write this, of course.
 */
std::unique_ptr<koopa_ast::Function>
translate_func_def_c_ast(const c_ast::FuncDefAST &func_def) {
  auto ret = std::make_unique<koopa_ast::Function>();

  // Get the function name
  ret->name = "@" + func_def.ident;

  // Get the type
  auto *type =
      dynamic_cast<const c_ast::FuncTypeAST *>(func_def.func_type.get());
  if (!type) {
    throw std::runtime_error("ir_builder error: FuncDefAST expects "
                             "FuncTypeAST at param `func_type`");
  }

  ret->type = translate_func_type_c_ast(*type);

  // Get the block
  auto *block = dynamic_cast<const c_ast::BlockAST *>(func_def.block.get());
  if (!block) {
    throw std::runtime_error(
        "ir_builder error: FuncDefAST expects BlockAST at param `block`");
  }

  // HACK: We only have one block for now.
  ret->basicblocks.push_back(translate_block_c_ast(*block, "%entry"));

  return ret;
}

/**
 * Converting a CompUnitAST in C to a program in koopa.
 */
std::unique_ptr<koopa_ast::Program>
translate_comp_unit_c_ast(const c_ast::CompUnitAST &comp_unit) {
  auto ret = std::make_unique<koopa_ast::Program>();

  auto *func_def =
      dynamic_cast<const c_ast::FuncDefAST *>(comp_unit.func_def.get());
  if (!func_def) {
    throw std::runtime_error(
        "ir_builder error: CompUnitAST expects FuncDefAST at param `func_def`");
  }

  ret->functions.push_back(translate_func_def_c_ast(*func_def));

  return ret;
}

/*
 * Exposed API for converting from C AST to Koopa Representation
 */
std::unique_ptr<koopa_ast::Program>
convert_to_custom_koopa_from_c_reps(std::unique_ptr<c_ast::BaseAST> ast) {
  auto *comp_unit = dynamic_cast<const c_ast::CompUnitAST *>(ast.get());

  if (!comp_unit) {
    throw std::runtime_error(
        "ir_builder error: Expects to have a C AST with root of CompUnitAST");
  }

  return translate_comp_unit_c_ast(*comp_unit);
}

File: rust/src/ir_builder.rs
use koopa::ir::{
    BasicBlock, BinaryOp, Function, FunctionData, Program, Value,
    builder::{BasicBlockBuilder, LocalInstBuilder, ValueBuilder},
};

use crate::c_ast::{
    AddExp, AddOp, Block, CompUnit, Exp, FuncDef, MulExp, MulOp, PrimaryExp, Stmt, UnaryExp,
    UnaryOp,
};

pub struct LowerCtx {
    pub program: Program,
    cur_bb: Option<BasicBlock>,
}

impl LowerCtx {
    pub fn new() -> Self {
        Self {
            program: Program::new(),
            cur_bb: None,
        }
    }

    pub fn build_program_from_comp_unit(&mut self, comp_unit: CompUnit) {
        self.lower_func_def(&comp_unit.func_def);
    }

    pub fn lower_func_def(&mut self, func_def: &FuncDef) {
        let func_data = self.func_header(func_def);
        let func = self.program.new_func(func_data);
        self.lower_function_body(func, &func_def.block);
    }

    pub fn lower_function_body(&mut self, func: Function, block: &Block) {
        let f: &mut FunctionData = self.program.func_mut(func);

        let entry: BasicBlock = f.dfg_mut().new_bb().basic_block(Some("%entry".into()));
        f.layout_mut()
            .bbs_mut()
            .push_key_back(entry)
            .expect("We should only have one basic block for now.");

        self.cur_bb = Some(entry);

        self.lower_block(func, block);

        self.cur_bb = None;
    }

    pub fn lower_block(&mut self, func: Function, block: &Block) {
        self.lower_stmt(func, &block.stmt);
    }

    pub fn lower_stmt(&mut self, func: Function, stmt: &Stmt) {
        let return_val = self.lower_exp(func, &stmt.exp);
        self.emit_return_inst(func, return_val);
    }

    pub fn lower_exp(&mut self, func: Function, exp: &Exp) -> Value {
        self.lower_add_exp(func, &exp.add_exp)
    }

    pub fn lower_primary_exp(&mut self, func: Function, p_exp: &PrimaryExp) -> Value {
        match p_exp {
            PrimaryExp::Exp { exp } => self.lower_exp(func, exp),
            PrimaryExp::Number { num } => self.construct_i32(func, *num),
        }
    }

    pub fn lower_unary_exp(&mut self, func: Function, unary_exp: &UnaryExp) -> Value {
        match unary_exp {
            UnaryExp::Primary { primary_exp } => self.lower_primary_exp(func, primary_exp),
            UnaryExp::OpUnary {
                unary_op,
                unary_exp,
            } => {
                let u_exp_val = self.lower_unary_exp(func, unary_exp);
                match unary_op {
                    UnaryOp::PLUS => u_exp_val,
                    UnaryOp::MINUS => {
                        let zero = self.construct_i32(func, 0);
                        self.emit_binary(func, BinaryOp::Sub, zero, u_exp_val)
                    }
                    UnaryOp::BANG => {
                        let zero = self.construct_i32(func, 0);
                        self.emit_binary(func, BinaryOp::Eq, u_exp_val, zero)
                    }
                    UnaryOp::TILDE => {
                        let all_one = self.construct_i32(func, -1);
                        self.emit_binary(func, BinaryOp::Xor, u_exp_val, all_one)
                    }
                }
            }
        }
    }

    pub fn lower_mul_exp(&mut self, func: Function, mul_exp: &MulExp) -> Value {
        match mul_exp {
            MulExp::Unary { unary_exp } => self.lower_unary_exp(func, unary_exp),
            MulExp::MulOpUnary {
                mul_exp,
                op,
                unary_exp,
            } => {
                let mul_exp_val = self.lower_mul_exp(func, mul_exp);
                let unary_exp_val = self.lower_unary_exp(func, unary_exp);
                match op {
                    MulOp::STAR => {
                        self.emit_binary(func, BinaryOp::Mul, mul_exp_val, unary_exp_val)
                    }
                    MulOp::SLASH => {
                        self.emit_binary(func, BinaryOp::Div, mul_exp_val, unary_exp_val)
                    }
                    MulOp::PERCENT => {
                        self.emit_binary(func, BinaryOp::Mod, mul_exp_val, unary_exp_val)
                    }
                }
            }
        }
    }

    pub fn lower_add_exp(&mut self, func: Function, add_exp: &AddExp) -> Value {
        match add_exp {
            AddExp::Mul { mul_exp } => self.lower_mul_exp(func, mul_exp),
            AddExp::AddOpMul {
                add_exp,
                op,
                mul_exp,
            } => {
                let add_exp_val = self.lower_add_exp(func, add_exp);
                let mul_exp_val = self.lower_mul_exp(func, mul_exp);
                match op {
                    AddOp::PLUS => self.emit_binary(func, BinaryOp::Add, add_exp_val, mul_exp_val),
                    AddOp::MINUS => self.emit_binary(func, BinaryOp::Sub, add_exp_val, mul_exp_val),
                }
            }
        }
    }

    pub fn emit_return_inst(&mut self, func: Function, return_val: Value) {
        let f = self.program.func_mut(func);
        let ret_inst = f.dfg_mut().new_value().ret(Some(return_val));
        self.push_inst(func, ret_inst);
    }

    pub fn emit_binary(
        &mut self,
        func: Function,
        binary_op: BinaryOp,
        lhs: Value,
        rhs: Value,
    ) -> Value {
        let f = self.program.func_mut(func);
        let binary_inst = f.dfg_mut().new_value().binary(binary_op, lhs, rhs);
        self.push_inst(func, binary_inst);
        binary_inst
    }

    pub fn construct_i32(&mut self, func: Function, n: i32) -> Value {
        let f = self.program.func_mut(func);
        f.dfg_mut().new_value().integer(n)
    }

    fn func_header(&self, func_def: &FuncDef) -> FunctionData {
        let name = if func_def.ident.starts_with('@') {
            func_def.ident.clone()
        } else {
            format!("@{}", func_def.ident)
        };
        FunctionData::new(name, Vec::new(), func_def.func_type.to_koopa_type())
    }

    fn push_inst(&mut self, func: Function, inst: Value) {
        let bb = self.cur_bb.expect("no current block set.");
        let f = self.program.func_mut(func);
        f.layout_mut()
            .bb_mut(bb)
            .insts_mut()
            .push_key_back(inst)
            .expect("Duplicated instruction node.");
    }
}

impl Default for LowerCtx {
    fn default() -> Self {
        Self::new()
    }
}
